Sources/NetworkProtection/Recovery/FailureRecoveryHandler.swift

```diff
@@ -0,0 +1,198 @@
+    public enum ServerHealth {
+        case healthy
+        case unhealthy
+    }
+
+    case started
+    case completed(ServerHealth)
+    case failed(_ error: Error)
+}
+
+protocol FailureRecoveryHandling {
+    func attemptRecovery(
+        to lastConnectedServer: NetworkProtectionServer,
+        includedRoutes: [IPAddressRange],
+        excludedRoutes: [IPAddressRange],
+        isKillSwitchEnabled: Bool,
+        updateConfig: @escaping (NetworkProtectionDeviceManagement.GenerateTunnelConfigResult) async throws -> Void
+    ) async
+
+    func stop() async
+}
+
+enum FailureRecoveryError: Error {
+    case noRecoveryNecessary
+}
+
+actor FailureRecoveryHandler: FailureRecoveryHandling {
+
+    struct RetryConfig {
+        let times: Int
+        let initialDelay: TimeInterval
+        let maxDelay: TimeInterval
+        let factor: Double
+
+        static var `default` = RetryConfig(
+            times: 5,
+            initialDelay: .seconds(30),
+            maxDelay: .minutes(5),
+            factor: 2.0
+        )
+    }
+
+    private let deviceManager: NetworkProtectionDeviceManagement
+    private weak var reassertingControl: Reasserting?
+    private let retryConfig: RetryConfig
+    private let eventHandler: (FailureRecoveryStep) -> Void
+
+    private var task: Task<Void, Never>? {
+        willSet {
+            task?.cancel()
+        }
+    }
+
+    init(deviceManager: NetworkProtectionDeviceManagement, reassertingControl: Reasserting, retryConfig: RetryConfig = .default, eventHandler: @escaping (FailureRecoveryStep) -> Void) {
+        self.deviceManager = deviceManager
+        self.reassertingControl = reassertingControl
+        self.retryConfig = retryConfig
+        self.eventHandler = eventHandler
+    }
+
+    func attemptRecovery(
+        to lastConnectedServer: NetworkProtectionServer,
+        includedRoutes: [IPAddressRange],
+        excludedRoutes: [IPAddressRange],
+        isKillSwitchEnabled: Bool,
+        updateConfig: @escaping (NetworkProtectionDeviceManagement.GenerateTunnelConfigResult) async throws -> Void
+    ) async {
+        await incrementalPeriodicChecks(retryConfig) { [weak self] in
+            guard let self else { return }
+            await reassertingControl?.startReasserting()
+            eventHandler(.started)
+            do {
+                let result = try await makeRecoveryAttempt(
+                    to: lastConnectedServer,
+                    includedRoutes: includedRoutes,
+                    excludedRoutes: excludedRoutes,
+                    isKillSwitchEnabled: isKillSwitchEnabled
+                )
+                try await updateConfig(result)
+                eventHandler(.completed(.unhealthy))
+                await reassertingControl?.stopReasserting()
+            } catch {
+                switch error {
+                case FailureRecoveryError.noRecoveryNecessary:
+                    await reassertingControl?.stopReasserting()
```

Comment:
This fixes the bug I mentioned yesterday.

Reply:
For this type of thing it's really important to ensure calls are always balanced.  If a method starts, it also should stop on every exit.  The `default` error case doesn't stop reasserting again.

The reason why this is extremely important is because otherwise you can find issues like this one, where reasserting starts but never again stops.

I see two possible solutions to this:
1. Add this right after starting the reasserting:
```swift
defer {
    await reassertingControl?.stopReasserting()
}
```
2. Since we're recovering through all the duration of the checks, why not start reasserting right before calling `incrementalPeriodicChecks` and stop reasserting right after that call?  I guess this could be an issue if something else also stops reassertions, though... so option 1 may be better.